src/main/java/org/spongepowered/common/mixin/core/world/MixinWorldServer.java

```diff
@@ -735,6 +740,31 @@ public void onUpdateTick(Block block, net.minecraft.world.World worldIn, BlockPo
 
     @Redirect(method = "addBlockEvent", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/WorldServer$ServerBlockEventList;add(Ljava/lang/Object;)Z", remap = false))
     public boolean onAddBlockEvent(WorldServer.ServerBlockEventList list, Object obj, BlockPos pos, Block blockIn, int eventId, int eventParam) {
+        // We fire a Pre event to make sure our captures do not get stuck in a loop.
+        // This is very common with pistons as they add block events while blocks are being notified.
+        if (blockIn instanceof BlockPistonBase) { // if it's a piston, try to simulate it
+            final IBlockState blockstate = this.getBlockState(pos);
+            EnumFacing direction = (EnumFacing) blockstate.getValue(BlockDirectional.FACING);
+
+            BlockPistonStructureHelper movedBlocks = new BlockPistonStructureHelper(this$, pos, direction, eventId == 0);
+            List<BlockPos> blocks = new ArrayList<BlockPos>();
+            blocks.addAll(movedBlocks.getBlocksToDestroy());
+            blocks.addAll(movedBlocks.getBlocksToMove());
+
+            List<Location<org.spongepowered.api.world.World>> locations = new ArrayList<Location<org.spongepowered.api.world.World>>();
+            for (BlockPos block : blocks) {
+                locations.add(new Location<>((org.spongepowered.api.world.World) this$, block.getX(), block.getY(), block.getZ()));
```

Comment:
You're not filtering block positions that have already been added, since there can be multiple duplications of locations being added.